[flaky test] test_copy_from_with_fail_fast_property_can_kill_all_nodes_with_failure_from_single_node

Disable temporarily, #11957

```
## FAILURE: Test test_copy_from_with_fail_fast_property_can_kill_all_nodes_with_failure_from_single_node(io.crate.integrationtests.CopyIntegrationTest)
java.lang.AssertionError: ## node: node_s0
{befb5a76-9b25-44ec-9a1a-725f67b22cf4=Task{id=befb5a76-9b25-44ec-9a1a-725f67b22cf4, tasks=[CollectTask{id=0, sharedContexts=SharedShardContexts{allocatedShards=[]}, consumer=DistributingConsumer{jobId=befb5a76-9b25-44ec-9a1a-725f67b22cf4, targetPhaseId=1, inputId=0, bucketIdx=0, downstreams=[Downstream{T73WU0yCRLOWImILnk_JKw', needsMoreData=true}]}, searchContexts=[], batchIterator=java.util.concurrent.CompletableFuture@25b5115f[Completed normally], finished=false}], closed=true}}

	at __randomizedtesting.SeedInfo.seed([9D69C4EF2BA53718:BA08338F368051F5]:0)
	at io.crate.integrationtests.SQLIntegrationTestCase.assertNoTasksAreLeftOpen(SQLIntegrationTestCase.java:320)
	at jdk.internal.reflect.GeneratedMethodAccessor99.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:1004)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:299)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:293)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.AssertionError: 
Expected: is <0>
     but: was <1>
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
	at org.junit.Assert.assertThat(Assert.java:964)
	at org.junit.Assert.assertThat(Assert.java:930)
	at io.crate.integrationtests.SQLIntegrationTestCase.lambda$assertNoTasksAreLeftOpen$0(SQLIntegrationTestCase.java:277)
	at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:770)
	at io.crate.integrationtests.SQLIntegrationTestCase.assertNoTasksAreLeftOpen(SQLIntegrationTestCase.java:272)
	... 11 more
	Suppressed: java.lang.AssertionError: 
Expected: is <0>
     but: was <1>
		at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
		at org.junit.Assert.assertThat(Assert.java:964)
		at org.junit.Assert.assertThat(Assert.java:930)
		at io.crate.integrationtests.SQLIntegrationTestCase.lambda$assertNoTasksAreLeftOpen$0(SQLIntegrationTestCase.java:277)
		at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:758)
		... 12 more
	Suppressed: java.lang.AssertionError: 
Expected: is <0>
     but: was <1>
		at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
		at org.junit.Assert.assertThat(Assert.java:964)
		at org.junit.Assert.assertThat(Assert.java:930)
		at io.crate.integrationtests.SQLIntegrationTestCase.lambda$assertNoTasksAreLeftOpen$0(SQLIntegrationTestCase.java:277)
		at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:758)
		... 12 more
	Suppressed: java.lang.AssertionError: 
Expected: is <0>
     but: was <1>
		at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
		at org.junit.Assert.assertThat(Assert.java:964)
		at org.junit.Assert.assertThat(Assert.java:930)
		at io.crate.integrationtests.SQLIntegrationTestCase.lambda$assertNoTasksAreLeftOpen$0(SQLIntegrationTestCase.java:277)
		at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:758)
		... 12 more
	Suppressed: java.lang.AssertionError: 
Expected: is <0>
     but: was <1>
		at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
		at org.junit.Assert.assertThat(Assert.java:964)
		at org.junit.Assert.assertThat(Assert.java:930)
		at io.crate.integrationtests.SQLIntegrationTestCase.lambda$assertNoTasksAreLeftOpen$0(SQLIntegrationTestCase.java:277)
		at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:758)
		... 12 more
	Suppressed: java.lang.AssertionError: 
Expected: is <0>
     but: was <1>
		at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
		at org.junit.Assert.assertThat(Assert.java:964)
		at org.junit.Assert.assertThat(Assert.java:930)
		at io.crate.integrationtests.SQLIntegrationTestCase.lambda$assertNoTasksAreLeftOpen$0(SQLIntegrationTestCase.java:277)
		at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:758)
		... 12 more
	Suppressed: java.lang.AssertionError: 
Expected: is <0>
     but: was <1>
		at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
		at org.junit.Assert.assertThat(Assert.java:964)
		at org.junit.Assert.assertThat(Assert.java:930)
		at io.crate.integrationtests.SQLIntegrationTestCase.lambda$assertNoTasksAreLeftOpen$0(SQLIntegrationTestCase.java:277)
		at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:758)
		... 12 more
	Suppressed: java.lang.AssertionError: 
Expected: is <0>
     but: was <1>
		at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
		at org.junit.Assert.assertThat(Assert.java:964)
		at org.junit.Assert.assertThat(Assert.java:930)
		at io.crate.integrationtests.SQLIntegrationTestCase.lambda$assertNoTasksAreLeftOpen$0(SQLIntegrationTestCase.java:277)
		at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:758)
		... 12 more
	Suppressed: java.lang.AssertionError: 
Expected: is <0>
     but: was <1>
		at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
		at org.junit.Assert.assertThat(Assert.java:964)
		at org.junit.Assert.assertThat(Assert.java:930)
		at io.crate.integrationtests.SQLIntegrationTestCase.lambda$assertNoTasksAreLeftOpen$0(SQLIntegrationTestCase.java:277)
		at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:758)
		... 12 more
	Suppressed: java.lang.AssertionError: 
Expected: is <0>
     but: was <1>
		at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
		at org.junit.Assert.assertThat(Assert.java:964)
		at org.junit.Assert.assertThat(Assert.java:930)
		at io.crate.integrationtests.SQLIntegrationTestCase.lambda$assertNoTasksAreLeftOpen$0(SQLIntegrationTestCase.java:277)
		at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:758)
		... 12 more
	Suppressed: java.lang.AssertionError: 
Expected: is <0>
     but: was <1>
		at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
		at org.junit.Assert.assertThat(Assert.java:964)
		at org.junit.Assert.assertThat(Assert.java:930)
		at io.crate.integrationtests.SQLIntegrationTestCase.lambda$assertNoTasksAreLeftOpen$0(SQLIntegrationTestCase.java:277)
		at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:758)
		... 12 more
	Suppressed: java.lang.AssertionError: 
Expected: is <0>
     but: was <1>
		at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
		at org.junit.Assert.assertThat(Assert.java:964)
		at org.junit.Assert.assertThat(Assert.java:930)
		at io.crate.integrationtests.SQLIntegrationTestCase.lambda$assertNoTasksAreLeftOpen$0(SQLIntegrationTestCase.java:277)
		at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:758)
		... 12 more
	Suppressed: java.lang.AssertionError: 
Expected: is <0>
     but: was <1>
		at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
		at org.junit.Assert.assertThat(Assert.java:964)
		at org.junit.Assert.assertThat(Assert.java:930)
		at io.crate.integrationtests.SQLIntegrationTestCase.lambda$assertNoTasksAreLeftOpen$0(SQLIntegrationTestCase.java:277)
		at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:758)
		... 12 more
	Suppressed: java.lang.AssertionError: 
Expected: is <0>
     but: was <1>
		at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
[2021-12-06T17:27:58,979][INFO ][i.c.i.CopyIntegrationTest] [[Time-limited test]] before test

[2021-12-06T17:27:58,979][INFO ][i.c.i.CopyIntegrationTest] [[Time-limited test]] [CopyIntegrationTest#test_copy_from_with_fail_fast_property_can_kill_all_nodes_with_failure_from_single_node]: setting up test

[2021-12-06T17:27:58,982][INFO ][o.e.c.m.MetadataIndexTemplateService] [[cratedb[node_s0][masterService#updateTask][T#1]]] adding template [random_index_template] for index patterns [*]

[2021-12-06T17:27:58,994][INFO ][i.c.i.CopyIntegrationTest] [[Time-limited test]] [CopyIntegrationTest#test_copy_from_with_fail_fast_property_can_kill_all_nodes_with_failure_from_single_node]: all set up test

[2021-12-06T17:27:58,997][INFO ][i.c.p.s.SetSessionPlan   ] [[cratedb[node_s0][netty-worker][T#1]]] SET SESSION STATEMENT WILL BE IGNORED: extra_float_digits

[2021-12-06T17:27:58,998][INFO ][i.c.p.s.SetSessionPlan   ] [[cratedb[node_s0][netty-worker][T#1]]] SET SESSION STATEMENT WILL BE IGNORED: application_name

[2021-12-06T17:27:59,004][INFO ][o.e.c.m.MetadataCreateIndexService] [[cratedb[node_s0][masterService#updateTask][T#1]]] [mdkznwkxkckv.tbl] creating index, cause [api], templates [random_index_template], shards [4]/[0], mappings [default]

[2021-12-06T17:27:59,004][INFO ][o.e.c.r.a.AllocationService] [[cratedb[node_s0][masterService#updateTask][T#1]]] updating number_of_replicas to [1] for indices [mdkznwkxkckv.tbl]

[2021-12-06T17:27:59,190][INFO ][o.e.c.r.a.AllocationService] [[cratedb[node_s0][masterService#updateTask][T#1]]] Cluster health status changed from [YELLOW] to [GREEN] (reason: [shards started [[mdkznwkxkckv.tbl][1]]]).

[2021-12-06T17:27:59,204][INFO ][i.c.p.s.SetSessionPlan   ] [[cratedb[node_s1][netty-worker][T#3]]] SET SESSION STATEMENT WILL BE IGNORED: extra_float_digits

[2021-12-06T17:27:59,205][INFO ][i.c.p.s.SetSessionPlan   ] [[cratedb[node_s1][netty-worker][T#3]]] SET SESSION STATEMENT WILL BE IGNORED: application_name

[2021-12-06T17:27:59,260][INFO ][i.c.p.s.SetSessionPlan   ] [[cratedb[node_sc2][netty-worker][T#3]]] SET SESSION STATEMENT WILL BE IGNORED: extra_float_digits

[2021-12-06T17:27:59,261][INFO ][i.c.p.s.SetSessionPlan   ] [[cratedb[node_sc2][netty-worker][T#3]]] SET SESSION STATEMENT WILL BE IGNORED: application_name

Error: 2-06T17:28:14,743][ERROR][i.c.t.SQLTransportExecutor] [[Time-limited test]] Error executing stmt=copy tbl from ? with (shared=true, fail_fast=true, bulk_size=2) return summary args=[file:///tmp/junit3437157713325476380/junit2543163352629364392/*] error=org.postgresql.util.PSQLException: ERROR: Job killed. NODE: node_s0

[URI: file:///tmp/junit3437157713325476380/junit2543163352629364392/d1c6c323-9c63-4b38-8ef7-2ffede3b729d, ERRORS: {failed to parse field [a] of type [integer]={count=1, line_numbers=[2]}}]

[2021-12-06T17:28:15,311][INFO ][i.c.p.s.SetSessionPlan   ] [[cratedb[node_s0][netty-worker][T#1]]] SET SESSION STATEMENT WILL BE IGNORED: extra_float_digits

[2021-12-06T17:28:15,312][INFO ][i.c.p.s.SetSessionPlan   ] [[cratedb[node_s0][netty-worker][T#1]]] SET SESSION STATEMENT WILL BE IGNORED: application_name

[2021-12-06T17:28:15,332][INFO ][i.c.p.s.SetSessionPlan   ] [[cratedb[node_s0][netty-worker][T#1]]] SET SESSION STATEMENT WILL BE IGNORED: extra_float_digits

[2021-12-06T17:28:15,334][INFO ][i.c.p.s.SetSessionPlan   ] [[cratedb[node_s0][netty-worker][T#1]]] SET SESSION STATEMENT WILL BE IGNORED: application_name

[2021-12-06T17:28:17,810][INFO ][i.c.p.s.SetSessionPlan   ] [[cratedb[node_s1][netty-worker][T#1]]] SET SESSION STATEMENT WILL BE IGNORED: extra_float_digits

[2021-12-06T17:28:17,811][INFO ][i.c.p.s.SetSessionPlan   ] [[cratedb[node_s1][netty-worker][T#1]]] SET SESSION STATEMENT WILL BE IGNORED: application_name

Error: 2-06T17:28:17,842][ERROR][i.c.t.SQLTransportExecutor] [[Time-limited test]] Error executing stmt=copy tbl from ? with (shared=true, fail_fast=true, bulk_size=2) return summary args=[file:///tmp/junit3437157713325476380/junit11625390341564996199/*] error=org.postgresql.util.PSQLException: ERROR: Job killed. NODE: node_s1

[URI: file:///tmp/junit3437157713325476380/junit11625390341564996199/eb5dcb05-28ac-4706-bf92-ba6c99db4697, ERRORS: {failed to parse field [a] of type [integer]={count=1, line_numbers=[2]}}]

[2021-12-06T17:28:17,952][INFO ][i.c.p.s.SetSessionPlan   ] [[cratedb[node_sc2][netty-worker][T#1]]] SET SESSION STATEMENT WILL BE IGNORED: extra_float_digits

[2021-12-06T17:28:17,953][INFO ][i.c.p.s.SetSessionPlan   ] [[cratedb[node_sc2][netty-worker][T#1]]] SET SESSION STATEMENT WILL BE IGNORED: application_name

[2021-12-06T17:28:17,978][INFO ][i.c.p.s.SetSessionPlan   ] [[cratedb[node_s0][netty-worker][T#1]]] SET SESSION STATEMENT WILL BE IGNORED: extra_float_digits

[2021-12-06T17:28:17,979][INFO ][i.c.p.s.SetSessionPlan   ] [[cratedb[node_s0][netty-worker][T#1]]] SET SESSION STATEMENT WILL BE IGNORED: application_name

[2021-12-06T17:28:18,030][INFO ][i.c.p.s.SetSessionPlan   ] [[cratedb[node_s0][netty-worker][T#1]]] SET SESSION STATEMENT WILL BE IGNORED: extra_float_digits

[2021-12-06T17:28:18,037][INFO ][i.c.p.s.SetSessionPlan   ] [[cratedb[node_s0][netty-worker][T#1]]] SET SESSION STATEMENT WILL BE IGNORED: application_name

[2021-12-06T17:28:28,208][INFO ][i.c.i.CopyIntegrationTest] [[Time-limited test]] [CopyIntegrationTest#test_copy_from_with_fail_fast_property_can_kill_all_nodes_with_failure_from_single_node]: cleaning up after test

[2021-12-06T17:28:28,219][INFO ][o.e.c.m.MetadataDeleteIndexService] [[cratedb[node_s0][masterService#updateTask][T#1]]] [mdkznwkxkckv.tbl/j6JuQ3wOTHe2CP_di_IhDw] deleting index

[2021-12-06T17:28:28,260][INFO ][o.e.c.m.MetadataIndexTemplateService] [[cratedb[node_s0][masterService#updateTask][T#1]]] removing template [random_index_template]

[2021-12-06T17:28:38,273][INFO ][i.c.i.CopyIntegrationTest] [[Time-limited test]] [CopyIntegrationTest#test_copy_from_with_fail_fast_property_can_kill_all_nodes_with_failure_from_single_node]: cleaned up after test

[2021-12-06T17:28:38,273][INFO ][i.c.i.CopyIntegrationTest] [[Time-limited test]] after test

REPRODUCE WITH: ./gradlew :server:test -Dtests.seed=9D69C4EF2BA53718 -Dtests.class=io.crate.integrationtests.CopyIntegrationTest -Dtests.method="test_copy_from_with_fail_fast_property_can_kill_all_nodes_with_failure_from_single_node" -Dtests.locale=en-UG -Dtests.timezone=Universal

		at org.junit.Assert.assertThat(Assert.java:964)
		at org.junit.Assert.assertThat(Assert.java:930)
		at io.crate.integrationtests.SQLIntegrationTestCase.lambda$assertNoTasksAreLeftOpen$0(SQLIntegrationTestCase.java:277)
		at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:758)
		... 12 more

CopyIntegrationTest > test_copy_from_with_fail_fast_property_can_kill_all_nodes_with_failure_from_single_node FAILED
```

--- a/server/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java
+++ b/server/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java
@@ -182,7 +182,7 @@ public class TransportShardUpsertAction extends TransportShardAction<ShardUpsert
                     throw Exceptions.toRuntimeException(e);
                 }
                 if (logger.isDebugEnabled()) {
-                    logger.debug("Failed to execute upsert shardId={} id={} error={}", request.shardId(), item.id(), e);
+                    logger.debug("Failed to execute upsert on nodeName={}, shardId={} id={} error={}", clusterService.localNode().getName(), request.shardId(), item.id(), e);
                 }
 
                 // *mark* the item as failed by setting the source to null

--- a/server/src/test/java/io/crate/execution/engine/indexing/UpsertResultsTest.java
+++ b/server/src/test/java/io/crate/execution/engine/indexing/UpsertResultsTest.java
@@ -24,8 +24,6 @@ package io.crate.execution.engine.indexing;
 import static org.hamcrest.CoreMatchers.containsString;
 import static org.hamcrest.CoreMatchers.is;
 import static org.hamcrest.MatcherAssert.assertThat;
-import static org.junit.jupiter.api.Assertions.assertEquals;
-import static org.junit.jupiter.api.Assertions.assertTrue;
 
 import org.junit.Test;
 
@@ -94,5 +92,12 @@ public class UpsertResultsTest {
                     [URI: file:///t2.json, ERRORS: {failed to parse={count=2, line_numbers=[4, 5]}}]"""
             )
         );
+    }
+
+    @Test
+    public void testResultToFailureSuccessMessages() {
+        UpsertResults upsertResults = new UpsertResults();
+        upsertResults.addResult("file:///t5.json", null, 5);
+        assertThat(UpsertResults.resultsToFailure(upsertResults).getMessage(), is("Job killed. \n[URI: file:///t5.json, ERRORS: {}]"));
     }
 }

--- a/server/src/test/java/io/crate/integrationtests/CopyFromFailFastITest.java
+++ b/server/src/test/java/io/crate/integrationtests/CopyFromFailFastITest.java
@@ -1,0 +1,273 @@
+/*
+ * Licensed to Crate.io GmbH ("Crate") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.integrationtests;
+
+import static com.carrotsearch.randomizedtesting.RandomizedTest.newTempDir;
+import static io.crate.integrationtests.CopyIntegrationTest.tmpFileWithLines;
+import static io.crate.testing.Asserts.assertThrowsMatches;
+import static org.hamcrest.Matchers.lessThan;
+import static org.hamcrest.core.Is.is;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.cluster.routing.OperationRouting;
+import org.elasticsearch.common.logging.Loggers;
+import org.elasticsearch.test.ESIntegTestCase;
+import org.elasticsearch.test.MockLogAppender;
+import org.elasticsearch.test.junit.annotations.TestLogging;
+import org.junit.After;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+import com.carrotsearch.randomizedtesting.LifecycleScope;
+
+import io.crate.action.sql.SQLOperations;
+import io.crate.exceptions.JobKilledException;
+import io.crate.testing.SQLTransportExecutor;
+import io.crate.testing.UseJdbc;
+
+@ESIntegTestCase.ClusterScope(numDataNodes = 0, numClientNodes = 0, supportsDedicatedMasters = false)
+public class CopyFromFailFastITest extends SQLIntegrationTestCase {
+
+    @Rule
+    public TemporaryFolder folder = new TemporaryFolder();
+
+    @After
+    public void resetSettings() {
+        execute("reset global " +
+                "overload_protection.dml.initial_concurrency, " +
+                "overload_protection.dml.max_concurrency," +
+                "overload_protection.dml.queue_size");
+    }
+
+    @UseJdbc(0)
+    @Test
+    public void test_copy_from_with_fail_fast_with_single_node() throws Exception {
+        internalCluster().startNode();
+        internalCluster().ensureAtLeastNumDataNodes(1);
+
+        // a single uri with 'shared = true' implies that one node will be involved at all times.
+        Path tmpDir = newTempDir(LifecycleScope.TEST);
+        Path target = Files.createDirectories(tmpDir.resolve("target"));
+        tmpFileWithLines(Arrays.asList(("{\"a\":987654321},".repeat(10) + "{\"a\":\"fail here\"}," +
+                                        "{\"a\":123456789},".repeat(10)).split(",")),
+                         "data1.json",
+                         target);
+
+        execute("CREATE TABLE t (a int) CLUSTERED INTO 1 SHARDS");
+
+        assertThrowsMatches(
+            () -> execute(
+                "COPY t FROM ? WITH (bulk_size = 1, fail_fast = true)", // fail_fast = true
+                new Object[]{target.toUri() + "*"}),
+            JobKilledException.class,
+            "ERRORS: {failed to parse field"
+        );
+    }
+
+    @TestLogging("io.crate.execution.dml.upsert:DEBUG")
+    @Test
+    public void test_copy_from_with_fail_fast_with_write_error_on_non_handler_node() throws Exception {
+        internalCluster().startNode();
+        internalCluster().startNode();
+        internalCluster().ensureAtLeastNumDataNodes(2);
+
+        execute("set global overload_protection.dml.initial_concurrency = 2");
+        execute("set global overload_protection.dml.max_concurrency = 2");
+        execute("set global overload_protection.dml.queue_size = 2");
+
+        execute("CREATE TABLE doc.t (a INT PRIMARY KEY, b INT) CLUSTERED INTO 2 SHARDS WITH (number_of_replicas=0)");
+
+        execute("SELECT node['name'] FROM sys.shards WHERE table_name='t' ORDER BY id");
+        var nodeNameOfShard0 = (String) response.rows()[0][0];
+        var nodeNameOfShard1 = (String) response.rows()[1][0];
+
+        var clientProvider = new SQLTransportExecutor.ClientProvider() {
+
+            @Override
+            public Client client() {
+                return internalCluster().client(nodeNameOfShard0);
+            }
+
+            @Override
+            public String pgUrl() {
+                return null;
+            }
+
+            @Override
+            public SQLOperations sqlOperations() {
+                return internalCluster().getInstance(SQLOperations.class, nodeNameOfShard0);
+            }
+        };
+        var handlerNodeExecutor = new SQLTransportExecutor(clientProvider);
+
+        Path tmpDir = newTempDir(LifecycleScope.TEST);
+        Path target = Files.createDirectories(tmpDir.resolve("target"));
+        int numDocs = 100;
+
+        var indexMetadata = clusterService().state().getMetadata().index("t");
+        List<String> rows = new ArrayList<>();
+        for (int i = 0; i < numDocs; i++) {
+            String line;
+            // Ensure the failing doc will fail on shard id 1
+            if (i >= 20 && OperationRouting.generateShardId(indexMetadata, Integer.toString(i), null) == 1) {
+                line = "{\"a\":" + i + ", \"b\":\"fail here\"}";
+            } else {
+                line = "{\"a\":" + i + ", \"b\":" + i + "}";
+            }
+            rows.add(line);
+        }
+
+        tmpFileWithLines(rows, "data1.json", target);
+        assertExpectedLogMessages(
+            () -> assertThrowsMatches(
+                () -> handlerNodeExecutor.exec(
+                    "COPY doc.t FROM ? WITH (bulk_size = 1, fail_fast = true, shared= true)", // fail_fast = true
+                    new Object[]{target.toUri() + "*"}),
+                JobKilledException.class,
+                "failed to parse field [b] of type [integer]"
+            ),
+            new MockLogAppender.PatternSeenEventExcpectation(
+                "assert failure on node=" + nodeNameOfShard1,
+                "io.crate.execution.dml.upsert.TransportShardUpsertAction",
+                Level.DEBUG,
+                "Failed to execute upsert on nodeName="+nodeNameOfShard1 + ".*")
+        );
+
+        execute("SELECT COUNT(*) FROM doc.t");
+        assertThat((long) response.rows()[0][0], lessThan(50L));
+    }
+
+    @Test
+    public void test_copy_from_fail_fast_with_passing_uris_only() throws IOException {
+
+        Path tmpDir = newTempDir(LifecycleScope.TEST);
+        Path target = Files.createDirectories(tmpDir.resolve("target"));
+        tmpFileWithLines(Collections.singletonList(("{\"a\":123}")),
+                         "passing1.json",
+                         target);
+        tmpFileWithLines(Arrays.asList(("{\"a\":123},".repeat(2).split(","))),
+                         "passing2.json",
+                         target);
+        tmpFileWithLines(Arrays.asList(("{\"a\":123},".repeat(3).split(","))),
+                         "passing3.json",
+                         target);
+
+        execute("CREATE TABLE t (a int)");
+
+        execute("COPY t FROM ? WITH (fail_fast = true, shared = false)", new Object[]{target.toUri().toString() + "*"});
+        refresh();
+        execute("select * from t");
+        assertThat(response.rowCount(), is(6L * cluster().numDataNodes()));
+
+        execute("COPY t FROM ? WITH (fail_fast = true, shared = false) return summary", new Object[]{target.toUri().toString() + "*"});
+        refresh();
+        execute("select * from t");
+        assertThat(response.rowCount(), is(6L * cluster().numDataNodes() * 2));
+    }
+
+    @Test
+    public void test_copy_from_fail_fast_with_passing_and_failing_uris() throws IOException {
+
+        Path tmpDir = newTempDir(LifecycleScope.TEST);
+        Path target = Files.createDirectories(tmpDir.resolve("target"));
+        tmpFileWithLines(Collections.singletonList(("{\"a\":123}")),
+                         "passing1.json",
+                         target);
+        tmpFileWithLines(Arrays.asList(("{\"a\":123},".repeat(2).split(","))),
+                         "passing2.json",
+                         target);
+        tmpFileWithLines(Arrays.asList(("{\"a\":123},".repeat(3).split(","))),
+                         "passing3.json",
+                         target);
+        tmpFileWithLines(Arrays.asList(("{\"a\":987654321},".repeat(2) + "{\"a\":\"fail here\"}," +
+                                        "{\"a\":123456789},".repeat(3)).split(",")),
+                         "failing1.json",
+                         target);
+        tmpFileWithLines(Arrays.asList(("{\"a\":987654321},".repeat(4) + "{\"b\":\"fail here\"}," +
+                                        "{\"a\":123456789},".repeat(5)).split(",")),
+                         "failing2.json",
+                         target);
+
+        String expectedMessage = "[URI: failing1.json, ERRORS: {failed to parse field [a] of type [integer]={count=1, line_numbers=[3]}}]" +
+                                 "[URI: failing2.json, ERRORS: {mapping set to strict, dynamic introduction of [b] within [default] is not allowed={count=1, line_numbers=[5]}}]" +
+                                 "[URI: passing1.json, ERRORS: {}][URI: passing2.json, ERRORS: {}][URI: passing3.json, ERRORS: {}]";
+
+        execute("CREATE TABLE t (a int)");
+
+        try {
+            execute("COPY t FROM ? WITH (fail_fast = true, shared = false)",
+                    new Object[]{target.toUri().toString() + "*"});
+        } catch (Exception e) {
+            String exceptionMessage = e.getMessage();
+            if (e.getMessage() == null || e.getMessage().isEmpty()) {
+                fail("expect a message from this exception");
+            }
+            exceptionMessage = exceptionMessage.replaceAll("file:///.*/target/", "").replaceAll("],","]");
+            exceptionMessage = Arrays.stream(exceptionMessage.split("\n")).filter(s -> s.startsWith("[URI:")).sorted().collect(
+                Collectors.joining());
+            assertThat(exceptionMessage, is(expectedMessage));
+        }
+
+        try {
+            execute("COPY t FROM ? WITH (fail_fast = true, shared = false) return summary",
+                    new Object[]{target.toUri().toString() + "*"});
+        } catch (Exception e) {
+            String exceptionMessage = e.getMessage();
+            if (e.getMessage() == null || e.getMessage().isEmpty()) {
+                fail("expect a message from this exception");
+            }
+            exceptionMessage = exceptionMessage.replaceAll("file:///.*/target/", "").replaceAll("],","]");
+            exceptionMessage = Arrays.stream(exceptionMessage.split("\n")).filter(s -> s.startsWith("[URI:")).sorted().collect(
+                Collectors.joining());
+            assertThat(exceptionMessage, is(expectedMessage));
+        }
+    }
+
+    private void assertExpectedLogMessages(Runnable command,
+                                           MockLogAppender.LoggingExpectation ... expectations) throws IllegalAccessException {
+        Logger testLogger = LogManager.getLogger("io.crate.execution.dml.upsert");
+        MockLogAppender appender = new MockLogAppender();
+        Loggers.addAppender(testLogger, appender);
+        try {
+            appender.start();
+            Arrays.stream(expectations).forEach(appender::addExpectation);
+            command.run();
+            appender.assertAllExpectationsMatched();
+        } finally {
+            Loggers.removeAppender(testLogger, appender);
+        }
+    }
+
+}

--- a/server/src/test/java/io/crate/integrationtests/CopyFromFailFastITest.java
+++ b/server/src/test/java/io/crate/integrationtests/CopyFromFailFastITest.java
@@ -24,8 +24,8 @@ package io.crate.integrationtests;
 import static com.carrotsearch.randomizedtesting.RandomizedTest.newTempDir;
 import static io.crate.integrationtests.CopyIntegrationTest.tmpFileWithLines;
 import static io.crate.testing.Asserts.assertThrowsMatches;
+import static org.hamcrest.Matchers.is;
 import static org.hamcrest.Matchers.lessThan;
-import static org.hamcrest.core.Is.is;
 
 import java.io.IOException;
 import java.nio.file.Files;
@@ -34,7 +34,6 @@ import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
-import java.util.stream.Collectors;
 
 import org.apache.logging.log4j.Level;
 import org.apache.logging.log4j.LogManager;
@@ -169,90 +168,95 @@ public class CopyFromFailFastITest extends SQLIntegrationTestCase {
         assertThat((long) response.rows()[0][0], lessThan(50L));
     }
 
+    @TestLogging("io.crate.execution.dml.upsert:DEBUG")
     @Test
-    public void test_copy_from_fail_fast_with_passing_uris_only() throws IOException {
+    public void test_copy_from_with_fail_fast_with_write_error_on_handler_node() throws Exception {
+        internalCluster().startNode();
+        internalCluster().startNode();
+        internalCluster().ensureAtLeastNumDataNodes(2);
+
+        execute("set global overload_protection.dml.initial_concurrency = 2");
+        execute("set global overload_protection.dml.max_concurrency = 2");
+        execute("set global overload_protection.dml.queue_size = 2");
+
+        execute("CREATE TABLE doc.t (a INT PRIMARY KEY, b INT) CLUSTERED INTO 2 SHARDS WITH (number_of_replicas=0)");
+
+        execute("SELECT node['name'] FROM sys.shards WHERE table_name='t' ORDER BY id");
+        var nodeNameOfShard0 = (String) response.rows()[0][0];
+
+        var clientProvider = new SQLTransportExecutor.ClientProvider() {
+
+            @Override
+            public Client client() {
+                return internalCluster().client(nodeNameOfShard0);
+            }
+
+            @Override
+            public String pgUrl() {
+                return null;
+            }
+
+            @Override
+            public SQLOperations sqlOperations() {
+                return internalCluster().getInstance(SQLOperations.class, nodeNameOfShard0);
+            }
+        };
+        var handlerNodeExecutor = new SQLTransportExecutor(clientProvider);
 
         Path tmpDir = newTempDir(LifecycleScope.TEST);
         Path target = Files.createDirectories(tmpDir.resolve("target"));
-        tmpFileWithLines(Collections.singletonList(("{\"a\":123}")),
-                         "passing1.json",
-                         target);
-        tmpFileWithLines(Arrays.asList(("{\"a\":123},".repeat(2).split(","))),
-                         "passing2.json",
-                         target);
-        tmpFileWithLines(Arrays.asList(("{\"a\":123},".repeat(3).split(","))),
-                         "passing3.json",
-                         target);
+        int numDocs = 100;
+
+        var indexMetadata = clusterService().state().getMetadata().index("t");
+        List<String> rows = new ArrayList<>();
+        for (int i = 0; i < numDocs; i++) {
+            String line;
+            // Ensure the failing doc will fail on shard id 0
+            if (i >= 20 && OperationRouting.generateShardId(indexMetadata, Integer.toString(i), null) == 0) {
+                line = "{\"a\":" + i + ", \"b\":\"fail here\"}";
+            } else {
+                line = "{\"a\":" + i + ", \"b\":" + i + "}";
+            }
+            rows.add(line);
+        }
+
+        tmpFileWithLines(rows, "data1.json", target);
+        assertExpectedLogMessages(
+            () -> assertThrowsMatches(
+                () -> handlerNodeExecutor.exec(
+                    "COPY doc.t FROM ? WITH (bulk_size = 1, fail_fast = true, shared= true)", // fail_fast = true
+                    new Object[]{target.toUri() + "*"}),
+                JobKilledException.class,
+                "failed to parse field [b] of type [integer]"
+            ),
+            new MockLogAppender.PatternSeenEventExcpectation(
+                "assert failure on node=" + nodeNameOfShard0,
+                "io.crate.execution.dml.upsert.TransportShardUpsertAction",
+                Level.DEBUG,
+                "Failed to execute upsert on nodeName="+nodeNameOfShard0 + ".*")
+        );
+
+        execute("SELECT COUNT(*) FROM doc.t");
+        assertThat((long) response.rows()[0][0], lessThan(50L));
+    }
+
+    @Test
+    public void test_copy_from_fail_fast_without_return_summary_without_write_errors() throws IOException {
+        // fail_fast = true and fail_fast = false that do not fail take different UpsertResultCollectors.
+        // fail_fast = false uses RowCountCollector while fail_fast = true uses newSummaryOnFailOrRowCountOnSuccessCollector
+        // and in the case that nothing fail, the collected summary will turn into rowCounts
+        internalCluster().startNode();
+
+        Path tmpDir = newTempDir(LifecycleScope.TEST);
+        Path target = Files.createDirectories(tmpDir.resolve("target"));
+        tmpFileWithLines(Collections.singletonList(("{\"a\":123}")), "passing1.json", target);
 
         execute("CREATE TABLE t (a int)");
 
         execute("COPY t FROM ? WITH (fail_fast = true, shared = false)", new Object[]{target.toUri().toString() + "*"});
         refresh();
         execute("select * from t");
-        assertThat(response.rowCount(), is(6L * cluster().numDataNodes()));
-
-        execute("COPY t FROM ? WITH (fail_fast = true, shared = false) return summary", new Object[]{target.toUri().toString() + "*"});
-        refresh();
-        execute("select * from t");
-        assertThat(response.rowCount(), is(6L * cluster().numDataNodes() * 2));
-    }
-
-    @Test
-    public void test_copy_from_fail_fast_with_passing_and_failing_uris() throws IOException {
-
-        Path tmpDir = newTempDir(LifecycleScope.TEST);
-        Path target = Files.createDirectories(tmpDir.resolve("target"));
-        tmpFileWithLines(Collections.singletonList(("{\"a\":123}")),
-                         "passing1.json",
-                         target);
-        tmpFileWithLines(Arrays.asList(("{\"a\":123},".repeat(2).split(","))),
-                         "passing2.json",
-                         target);
-        tmpFileWithLines(Arrays.asList(("{\"a\":123},".repeat(3).split(","))),
-                         "passing3.json",
-                         target);
-        tmpFileWithLines(Arrays.asList(("{\"a\":987654321},".repeat(2) + "{\"a\":\"fail here\"}," +
-                                        "{\"a\":123456789},".repeat(3)).split(",")),
-                         "failing1.json",
-                         target);
-        tmpFileWithLines(Arrays.asList(("{\"a\":987654321},".repeat(4) + "{\"b\":\"fail here\"}," +
-                                        "{\"a\":123456789},".repeat(5)).split(",")),
-                         "failing2.json",
-                         target);
-
-        String expectedMessage = "[URI: failing1.json, ERRORS: {failed to parse field [a] of type [integer]={count=1, line_numbers=[3]}}]" +
-                                 "[URI: failing2.json, ERRORS: {mapping set to strict, dynamic introduction of [b] within [default] is not allowed={count=1, line_numbers=[5]}}]" +
-                                 "[URI: passing1.json, ERRORS: {}][URI: passing2.json, ERRORS: {}][URI: passing3.json, ERRORS: {}]";
-
-        execute("CREATE TABLE t (a int)");
-
-        try {
-            execute("COPY t FROM ? WITH (fail_fast = true, shared = false)",
-                    new Object[]{target.toUri().toString() + "*"});
-        } catch (Exception e) {
-            String exceptionMessage = e.getMessage();
-            if (e.getMessage() == null || e.getMessage().isEmpty()) {
-                fail("expect a message from this exception");
-            }
-            exceptionMessage = exceptionMessage.replaceAll("file:///.*/target/", "").replaceAll("],","]");
-            exceptionMessage = Arrays.stream(exceptionMessage.split("\n")).filter(s -> s.startsWith("[URI:")).sorted().collect(
-                Collectors.joining());
-            assertThat(exceptionMessage, is(expectedMessage));
-        }
-
-        try {
-            execute("COPY t FROM ? WITH (fail_fast = true, shared = false) return summary",
-                    new Object[]{target.toUri().toString() + "*"});
-        } catch (Exception e) {
-            String exceptionMessage = e.getMessage();
-            if (e.getMessage() == null || e.getMessage().isEmpty()) {
-                fail("expect a message from this exception");
-            }
-            exceptionMessage = exceptionMessage.replaceAll("file:///.*/target/", "").replaceAll("],","]");
-            exceptionMessage = Arrays.stream(exceptionMessage.split("\n")).filter(s -> s.startsWith("[URI:")).sorted().collect(
-                Collectors.joining());
-            assertThat(exceptionMessage, is(expectedMessage));
-        }
+        assertThat(response.rowCount(), is((long) cluster().numDataNodes()));
     }
 
     private void assertExpectedLogMessages(Runnable command,

--- a/server/src/test/java/io/crate/integrationtests/CopyIntegrationTest.java
+++ b/server/src/test/java/io/crate/integrationtests/CopyIntegrationTest.java
@@ -21,38 +21,6 @@
 
 package io.crate.integrationtests;
 
-import com.carrotsearch.randomizedtesting.LifecycleScope;
-import io.crate.execution.engine.collect.files.FileReadingIterator;
-import io.crate.execution.engine.collect.sources.FileCollectSource;
-import io.crate.testing.SQLResponse;
-import io.crate.testing.UseJdbc;
-import org.elasticsearch.test.ESIntegTestCase;
-import org.elasticsearch.threadpool.ThreadPool;
-import org.junit.Rule;
-import org.junit.Test;
-import org.junit.rules.TemporaryFolder;
-
-import java.io.File;
-import java.io.FileOutputStream;
-import java.io.IOException;
-import java.io.OutputStreamWriter;
-import java.net.URI;
-import java.net.URISyntaxException;
-import java.nio.charset.StandardCharsets;
-import java.nio.file.DirectoryStream;
-import java.nio.file.Files;
-import java.nio.file.Path;
-import java.nio.file.Paths;
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Locale;
-import java.util.Map;
-import java.util.UUID;
-import java.util.stream.Collectors;
-
 import static com.carrotsearch.randomizedtesting.RandomizedTest.newTempDir;
 import static io.crate.protocols.postgres.PGErrorStatus.INTERNAL_ERROR;
 import static io.crate.testing.Asserts.assertThrowsMatches;
@@ -68,6 +36,35 @@ import static org.hamcrest.Matchers.notNullValue;
 import static org.hamcrest.Matchers.nullValue;
 import static org.hamcrest.Matchers.startsWith;
 import static org.hamcrest.core.Is.is;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.OutputStreamWriter;
+import java.net.URISyntaxException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.DirectoryStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.UUID;
+
+import org.elasticsearch.test.ESIntegTestCase;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+import com.carrotsearch.randomizedtesting.LifecycleScope;
+
+import io.crate.testing.SQLResponse;
+import io.crate.testing.UseJdbc;
 
 @ESIntegTestCase.ClusterScope(numDataNodes = 2)
 public class CopyIntegrationTest extends SQLHttpIntegrationTest {
@@ -86,6 +83,7 @@ public class CopyIntegrationTest extends SQLHttpIntegrationTest {
 
     public CopyIntegrationTest() throws URISyntaxException {
     }
+
 
     @Test
     public void testCopyFromFileWithJsonExtension() throws Exception {
@@ -643,14 +641,14 @@ public class CopyIntegrationTest extends SQLHttpIntegrationTest {
         return tmpFileWithLines(List.of(r1, r2, r3));
     }
 
-    private static Path tmpFileWithLines(Iterable<String> lines) throws IOException {
+    public static Path tmpFileWithLines(Iterable<String> lines) throws IOException {
         Path tmpDir = newTempDir(LifecycleScope.TEST);
         Path target = Files.createDirectories(tmpDir.resolve("target"));
         tmpFileWithLines(lines, "data.json", target);
         return Files.createSymbolicLink(tmpDir.resolve("link"), target);
     }
 
-    private static void tmpFileWithLines(Iterable<String> lines, String filename, Path target) throws IOException {
+    public static void tmpFileWithLines(Iterable<String> lines, String filename, Path target) throws IOException {
         File file = new File(target.toFile(), filename);
         Files.write(file.toPath(), lines, StandardCharsets.UTF_8);
     }
@@ -849,277 +847,5 @@ public class CopyIntegrationTest extends SQLHttpIntegrationTest {
         assertThat(printedTable(response.rows()), is(
             "1626188198073| 1625097600000\n"
         ));
-    }
-
-    @Test
-    public void test_copy_from_with_fail_fast_property_with_one_node_configuration() throws Exception {
-
-        // a single uri with 'shared = true' implies that one node will be involved at all times.
-
-        Path tmpDir = newTempDir(LifecycleScope.TEST);
-        Path target = Files.createDirectories(tmpDir.resolve("target"));
-        tmpFileWithLines(Arrays.asList(("{\"a\":987654321},".repeat(200) + "{\"a\":\"fail here\"}," +
-                                        "{\"a\":123456789},".repeat(200)).split(",")),
-                         "data1.json",
-                         target);
-
-        execute("CREATE TABLE t (a int)");
-        execute(
-            "COPY t FROM ? WITH (bulk_size = 1, fail_fast = false, shared = true)",
-            new Object[]{target.toUri().toString() + "*"});
-        refresh();
-        execute("select * from t");
-        assertThat(response.rowCount(), is(400L));
-
-        int maxTry = 100;
-        while (maxTry-- > 0) {
-            execute("drop table if exists t");
-            execute("CREATE TABLE t (a int)");
-            try {
-                execute(
-                    "COPY t FROM ? WITH (bulk_size = 1, fail_fast = true, shared = true)",
-                    new Object[]{target.toUri().toString() + "*"});
-                fail();
-            } catch (Exception e) {
-                String exceptionMessage = e.getMessage();
-                if (exceptionMessage == null || exceptionMessage.contains("failed to parse field") == false) {
-                    fail();
-                }
-            }
-            refresh();
-            execute("select * from t");
-            if (response.rowCount() < 400L) {
-                return;
-            }
-        }
-        fail();
-    }
-
-    @Test
-    public void test_copy_from_with_fail_fast_property_set_to_true_and_share_set_to_false() throws Exception {
-
-        // 'shared = false' will cause fail_fast to occur from all nodes
-
-        Path tmpDir = newTempDir(LifecycleScope.TEST);
-        Path target = Files.createDirectories(tmpDir.resolve("target"));
-        tmpFileWithLines(Arrays.asList(("{\"a\":987654321},".repeat(200) + "{\"a\":\"fail here\"}," +
-                                        "{\"a\":123456789},".repeat(200)).split(",")),
-                         "data1.json",
-                         target);
-
-        execute("CREATE TABLE t (a int)");
-        execute(
-            "COPY t FROM ? WITH (bulk_size = 1, fail_fast = false, shared = false)",
-            new Object[]{target.toUri().toString() + "*"});
-        refresh();
-        execute("select * from t");
-        assertThat(response.rowCount(), is(400L * cluster().numDataNodes()));
-
-        int maxTry = 100;
-        while (maxTry-- > 0) {
-            execute("drop table if exists t");
-            execute("CREATE TABLE t (a int)");
-            try {
-                execute(
-                    "COPY t FROM ? WITH (bulk_size = 1, fail_fast = true, shared = false)",
-                    new Object[]{target.toUri().toString() + "*"});
-                fail();
-            } catch (Exception e) {
-                String exceptionMessage = e.getMessage();
-                if (exceptionMessage == null || exceptionMessage.contains("failed to parse field") == false) {
-                    fail();
-                }
-            }
-            refresh();
-            execute("select * from t");
-            if (response.rowCount() < 400L * cluster().numDataNodes()) {
-                return;
-            }
-        }
-        fail();
-    }
-
-    private URI generateURIToBeConsumedByProvidedReaderNumber(Path target, List<String> data, int readerNumber, int numReaders) throws IOException {
-
-        int maxTry = 100;
-        URI uri;
-        while (maxTry-- > 0) {
-            File file = new File(target.toFile(), UUID.randomUUID().toString());
-            uri = FileReadingIterator.toURI(file.toPath().toUri().toString());
-            if (FileReadingIterator.moduloPredicateImpl(uri, readerNumber, numReaders)) {
-                Files.write(file.toPath(), data, StandardCharsets.UTF_8);
-                return uri;
-            }
-        }
-        return null;
-    }
-
-    @AwaitsFix(bugUrl = "https://github.com/crate/crate/issues/11956")
-    @Test
-    public void test_copy_from_with_fail_fast_property_can_kill_all_nodes_with_failure_from_single_node() throws Exception {
-
-        execute("create table tbl (a int)");
-        ensureGreen();
-        execute("select distinct node['name'], node['id'] from sys.shards where table_name = 'tbl'");
-        List<String> nodeNames = new ArrayList<>();
-        List<String> nodeIds = new ArrayList<>();
-        for (int i = 0; i < response.rowCount(); i++) {
-            nodeNames.add((String) response.rows()[i][0]);
-            nodeIds.add((String) response.rows()[i][1]);
-        }
-
-        for (int i = 0; i < nodeIds.size(); i++) {
-
-            // each node is assigned a readerNumber and based on URI.hashCode(), readerNumber, numReaders, the URIs will be assigned to a particular node to be consumed.
-            int readerNumber = FileCollectSource.getReaderNumber(nodeIds, nodeIds.get(i));
-
-            Path target = folder.newFolder().toPath();
-            URI failingUri = generateURIToBeConsumedByProvidedReaderNumber(
-                target,
-                Arrays.asList(("{\"a\":987654321}," + "{\"a\":\"fail here\"}").split(",")),
-                readerNumber, // making sure that every node gets a turn to consume failing uri.
-                nodeIds.size());
-            URI passingUri = generateURIToBeConsumedByProvidedReaderNumber(
-                target,
-                Arrays.asList("{\"a\":123456789},".repeat(10000).split(",")),
-                (readerNumber+1) % nodeIds.size(), // even with #nodes > 2, it does not matter who gets the passing uri.
-                nodeIds.size());
-            if (failingUri == null || passingUri == null) {
-                fail("two uris should be available to be copied from");
-            }
-            if (FileReadingIterator.moduloPredicateImpl(failingUri, readerNumber, nodeIds.size()) ==
-                FileReadingIterator.moduloPredicateImpl(passingUri, readerNumber, nodeIds.size())) {
-                fail("the two uris should not be consumed by the same node"); // the goal is to observe kill signals to spread across the nodes
-            }
-
-            String exceptionMessage = null;
-            try {
-                execute("copy tbl from ? with (shared=true, fail_fast=true, bulk_size=2) return summary",
-                        new Object[]{target.toUri().toString() + "*"});
-                fail();
-            } catch (Exception e) {
-                exceptionMessage = e.getMessage();
-                if (exceptionMessage == null || exceptionMessage.contains("failed to parse field") == false) {
-                    fail();
-                }
-            }
-
-            refresh();
-            waitUntilThreadPoolTasksFinished(ThreadPool.Names.SEARCH);
-
-            // bulk_size is 2 and with the failing URI containing 1 passing and 1 failing record, the 1 passing record must be inserted
-            execute("select count(*) from tbl where a = '987654321'");
-            assertEquals(1L, (long) response.rows()[0][0]);
-
-            // if the count is not less than 10000L, it is unclear whether fail_fast had any effect.
-            execute("select count(*) from tbl where a = '123456789'");
-            assertTrue((long) response.rows()[0][0] < 10000L);
-
-            // parse the exception message to retrieve the nodeName that observed the URI error and trigger the kill signals
-            int begin = exceptionMessage.indexOf("NODE: ");
-            if (begin == -1) {
-                fail("something went wrong with the expected exception message");
-            }
-            int end = exceptionMessage.indexOf("\n[URI");
-            if (end == -1) {
-                fail("something went wrong with the expected exception message");
-            }
-            String nodeName = exceptionMessage.substring(begin + "NODE: ".length(), end);
-
-            if (nodeName.equals(nodeNames.get(i)) == false) {
-                fail("nodeNames.get(i) should've been assigned to ingest the failing URI then should've triggered the failure");
-                // the purpose of this is to test that fail_fast can be triggered from handlerNode as well as non-handlerNodes
-                // since the handleNode will execute a CollectTask and a DownstreamRXTask whereas other data nodes will execute a CollectTask only
-            }
-
-            execute("delete from tbl");
-            refresh();
-        }
-    }
-
-    @Test
-    public void test_copy_from_fail_fast_with_passing_uris_only() throws IOException {
-
-        Path tmpDir = newTempDir(LifecycleScope.TEST);
-        Path target = Files.createDirectories(tmpDir.resolve("target"));
-        tmpFileWithLines(Collections.singletonList(("{\"a\":123}")),
-                         "passing1.json",
-                         target);
-        tmpFileWithLines(Arrays.asList(("{\"a\":123},".repeat(2).split(","))),
-                         "passing2.json",
-                         target);
-        tmpFileWithLines(Arrays.asList(("{\"a\":123},".repeat(3).split(","))),
-                         "passing3.json",
-                         target);
-
-        execute("CREATE TABLE t (a int)");
-
-        execute("COPY t FROM ? WITH (fail_fast = true, shared = false)", new Object[]{target.toUri().toString() + "*"});
-        refresh();
-        execute("select * from t");
-        assertThat(response.rowCount(), is(6L * cluster().numDataNodes()));
-
-        execute("COPY t FROM ? WITH (fail_fast = true, shared = false) return summary", new Object[]{target.toUri().toString() + "*"});
-        refresh();
-        execute("select * from t");
-        assertThat(response.rowCount(), is(6L * cluster().numDataNodes() * 2));
-    }
-
-    @Test
-    public void test_copy_from_fail_fast_with_passing_and_failing_uris() throws IOException {
-
-        Path tmpDir = newTempDir(LifecycleScope.TEST);
-        Path target = Files.createDirectories(tmpDir.resolve("target"));
-        tmpFileWithLines(Collections.singletonList(("{\"a\":123}")),
-                         "passing1.json",
-                         target);
-        tmpFileWithLines(Arrays.asList(("{\"a\":123},".repeat(2).split(","))),
-                         "passing2.json",
-                         target);
-        tmpFileWithLines(Arrays.asList(("{\"a\":123},".repeat(3).split(","))),
-                         "passing3.json",
-                         target);
-        tmpFileWithLines(Arrays.asList(("{\"a\":987654321},".repeat(2) + "{\"a\":\"fail here\"}," +
-                                        "{\"a\":123456789},".repeat(3)).split(",")),
-                         "failing1.json",
-                         target);
-        tmpFileWithLines(Arrays.asList(("{\"a\":987654321},".repeat(4) + "{\"b\":\"fail here\"}," +
-                                        "{\"a\":123456789},".repeat(5)).split(",")),
-                         "failing2.json",
-                         target);
-
-        String expectedMessage = "[URI: failing1.json, ERRORS: {failed to parse field [a] of type [integer]={count=1, line_numbers=[3]}}]" +
-                                 "[URI: failing2.json, ERRORS: {mapping set to strict, dynamic introduction of [b] within [default] is not allowed={count=1, line_numbers=[5]}}]" +
-                                 "[URI: passing1.json, ERRORS: {}][URI: passing2.json, ERRORS: {}][URI: passing3.json, ERRORS: {}]";
-
-        execute("CREATE TABLE t (a int)");
-
-        try {
-            execute("COPY t FROM ? WITH (fail_fast = true, shared = false)",
-                    new Object[]{target.toUri().toString() + "*"});
-        } catch (Exception e) {
-            String exceptionMessage = e.getMessage();
-            if (e.getMessage() == null || e.getMessage().isEmpty()) {
-                fail("expect a message from this exception");
-            }
-            exceptionMessage = exceptionMessage.replaceAll("file:///.*/target/", "").replaceAll("],","]");
-            exceptionMessage = Arrays.stream(exceptionMessage.split("\n")).filter(s -> s.startsWith("[URI:")).sorted().collect(
-                Collectors.joining());
-            assertThat(exceptionMessage, is(expectedMessage));
-        }
-
-        try {
-            execute("COPY t FROM ? WITH (fail_fast = true, shared = false) return summary",
-                    new Object[]{target.toUri().toString() + "*"});
-        } catch (Exception e) {
-            String exceptionMessage = e.getMessage();
-            if (e.getMessage() == null || e.getMessage().isEmpty()) {
-                fail("expect a message from this exception");
-            }
-            exceptionMessage = exceptionMessage.replaceAll("file:///.*/target/", "").replaceAll("],","]");
-            exceptionMessage = Arrays.stream(exceptionMessage.split("\n")).filter(s -> s.startsWith("[URI:")).sorted().collect(
-                Collectors.joining());
-            assertThat(exceptionMessage, is(expectedMessage));
-        }
     }
 }

